### ISSUE TITLE
allow using VOMS group for using local pilot in gWms

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#5090]
see instruction from Brian in this thread
https://hypernews.cern.ch/HyperNews/CMS/get/comp-ops/2671/1/1/1/1/1/1/2.html

**Comment by bbockelm** [comment]
Hm - not sure if this is going to do what you want.  Doesn't `user_fqan` refer only to the first of the VOMS attributes?

Rather, we should list all VOMS groups.

**Comment by belforte** [comment]
for sure this could still LARGELY benefit from a more clear documentation of how things are expected to work, which may lead to some code simplification as well. I find these comments indicative:
https://github.com/mmascher/CRABServer/blob/4863255a769cf6c4278a65b4d58648de965303a7/src/python/CMSGroupMapper.py#L21-L23
With all love and high respect for all who contributed... what does that mean ?

Sounds like the only hope to understand is to look up the code who uses this and check what is passed as argument.... ahem...  combine with the extensive usage of `def blah(*args,**kwargs)` and we always have code where suddently something is expected about presence,content and meaning of methods arguments which is not described and can only be verified going (at times a few levels) up in the call stack.  No wander at times even the best put in code that does not do what was expected.

**Comment by mmascher** [comment]
Hi @bbockelm,

the way things work in the client are that a user has to select if he wants to use one voms group using: `config.User.voGroup = 'group'`. This is saved in the oracle db so the TW can create a proxy with this group later on.

Therefore the only "valid" group is the first one returned by `voms-proxy-info -fqan`. Basically if a users does not use `config.User.voGroup = 'group'` then he says he is not going to use that group, and we should ignore that he can use local resources (IMHO).

I'll make an effort to improve the doc!

PS I have to reactivate github notifications, I am not getting these messages as emails..

**Comment by belforte** [comment]
Marco,
I think there is some misunderstanding. Please do not merge the change until this is sorted out.
Let's try to find a time where you and me can talk, if we can get Brian as well so much better, but
with 7 time zones, two newly borns and my usual things.. I am not sure we can get the 3-fold 
coincidence.

**Comment by bbockelm** [comment]
Hi,

I should be available tomorrow morning / early afternoon to discuss this.

I think processing *all* attributes is a fairly important aspect of the work, especially as an improvement to the old country-based VOMS matching.  The local user items have always assumed no action on the user's part to access the resources.  I don't want to give that up unless necessary.

Brian

**Comment by belforte** [comment]
ON second thought, let me try to recap and hopefully no meeting is needed,
other than maybe me and Marco (quick, in our morning and in italian).
The thing to keep in mind is that each proxy always "belongs to all groups",
but one group can be selected at proxy creation time to be primary.

1. I agree with Brian
2. in the end there not a right way and a wrong way, only what is more
  convenient for people/sites
3. we have two different use cases around
A) historical: site X (belgium e.g.) wants jobs to indicate a given group
   (becms e.g.) as PRIMARY group in VOMS FQAN, in order to grant some right
  (write access to local storage in that case).
B) the new one: site X wants CRAB to identify all users which belong to the
  same nationality so that gWms matches then to local pilots at the site,
  but the site does not care which primary group user have (at least
  Germans and Spain do not).
4. The topic so far refers to B.
5. Marco refers to A.

6. there is no deep reason why e.g. Belgium could not do with
  "becms is one of the listed groups, not necessarely the primary one",
  I guess that simply this way was easier for them to setup, or is the
  only one supported by there SRM, but that's all "accidental".


So A. will keep working as before, but for B we go for
"if one VOMS attribute matches one string in the site's local file,
we add the site name to the list in the cmsGroup classad"
And as soon as we add username as a VOMS attribute, we can handle it
in the same way.

If we wanted to be fully general and very precise, we should ask sites
to put in the local file not the simple /cms/dcms string (to identify one
group a DN may belong to) but create a language by which they can indicate
if they want a match with ANY group or with the primary FQAN, or whatever else.

I guess we do not want to do that, unless really forced by dire facts.




On 12/01/2015 09:52 PM, Brian Bockelman wrote:
> Hi,
>
> I should be available tomorrow morning / early afternoon to discuss this.
>
> I think processing /all/ attributes is a fairly important aspect of the work, especially as an
> improvement to the old country-based VOMS matching. The local user items have always assumed no
> action on the user's part to access the resources. I don't want to give that up unless necessary.
>
> Brian
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dmwm/CRABServer/issues/5090#issuecomment-161091770>.
>

-- 
Stefano Belforte - I.N.F.N.     tel   : +39 040 375-6261 (fax: 375-6258)
Area di Ricerca - Padriciano 99 tel mobile: +39 328 010 7327
34012 TRIESTE TS  - Italy       AIM: stefanobelforte

**Comment by bbockelm** [comment]
Hi Stefano,

Two few items:
1.  One slight correction.  We should advertise all VOMS groups in the CMSGroups attribute.  The matching between CMS group and what the site considers "local" is done by the glidein frontend (we just fixed this) and the startd.
   * This is done wrong in Marco's commit.  The local-users.txt should only map user name to the group named $SITE.  VOMS groups shouldn't be matched to this.  We are introducing other group names, such as `/cms/dcms`.  On the production side, we have group names like `HIG` and `TOP`.
   * The site will create `local-groups.txt` with the line `/cms/dcms`; the site name is implicitly added to that file.
2.  We want to make sure that we are not looking at the _VOMS group_, not the attribute (which includes `Role=user` or whatnot).  I think I might have said this incorrectly previously.

Brian

**Comment by mmascher** [comment]
Ok, sounds good. Let me write here if I understood:

In the siteconfs there are going to be two files, `local-groups.txt` (I missed this!) and `local-users.txt`, one containing usernames (`mmascher`) and one containing groups (`/cms/dcms`).

`CMSGroups` will still be a list of sites. The mechanism for `local-users.txt` will be the same: if a user is inside `SITE/local-users` we add SITE to the CMSGroups classad. For groups we are going to get all the groups (see later), and if at least one of the group is in `SITE/local-groups` we add SITE to CMSGroups.

About how to get groups, how is the best way? Naively I would call this:

```
[mmascher@lxplus0111 ~]$ voms-proxy-info -fqan
/cms/Role=NULL/Capability=NULL
/cms/integration/Role=NULL/Capability=NULL
```

and strip the `/Role=NULL/Capability=NULL` part.

**Comment by belforte** [comment]
OUch! Now I am confused.
Two quick questions:
1. so site maintains two files ? local-groups.txt and local-users.txt
  

2. I can't parse your point 1. below (one "not" too many ? is "whatnot" a word ? if so I do
not know it). Maybe an example will work better ?
e.g. this is from my proxy:
```
belforte@lxplus0091/~> voms-proxy-info -fqan
/cms/Role=NULL/Capability=NULL
/cms/ALARM/Role=NULL/Capability=NULL
/cms/TEAM/Role=NULL/Capability=NULL
/cms/dcms/Role=NULL/Capability=NULL
/cms/itcms/Role=NULL/Capability=NULL
```
I expect CRAB to put in JDL
`cmsGroup='/cms/dcms,/cmsitcms'`

and the FE will then match the job to sites who list any of those in their local-groups.txt

This appears to me a bit different from what Marco wrote :-(

while if there is "belforte" in local-users.txt for site T2_US_Nebraska,  CRAB will put in JDL
```
cmsGroups='/cms/dcms,/cms/itcms,T2_US_Nebraska'`

And we will need some rule to skip ALARM, TEAM etc. ...

I am sure you agree that this is not a "crisp specification" in they way Myron may like

**Comment by bbockelm** [comment]
Hi Stefano,

Your interpretation is incorrect - Marco is wrong.  All groups are listed in the CMSGroups attribute; matching of groups to sites is done by the frontend and startd.

The definition of a group is purposely loose - in the beginning, we defined only one way of determining group membership (an entry in `local-users.txt` for `SITE` means that you are in a group called `SITE`).

It was meant to be extended later on.  VOMS is one step.  WMAgent is now labelling workflows based on physics group.  I still would like to be able to list e-groups, as these have shown to be more popular within CMS than VOMS.

I don't think we'll need special rules for skipping ALARM, TEAM, etc - these only have effect if a site wants to prioritize these users.

**Comment by belforte** [comment]
what's the correct one then ?

**Comment by bbockelm** [comment]
Bah.  That's what I get for writing a comment before my first cup of coffee.

Stefano is _correct_; Marco is wrong.

I'll edit my original comment so future readers don't get confused.

**Comment by belforte** [comment]
I will try to add a bit of text to the twiki to summarize things,
hopefully will help me not to get confused again. Feel free to remind me, in case.

**Comment by bbockelm** [comment]
What's the twiki URL again?  I'll happily review things once you're done.

**Comment by mmascher** [comment]
There is an entry "site customized glideins" in the CRAB3 sw guide https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideCrab#Local_Submission

**Comment by belforte** [comment]
On 12/02/2015 03:22 PM, Brian Bockelman wrote:
> What's the twiki URL again?

finding it will be my first step. And a way to find all those twiki's...
I guess this:
https://twiki.cern.ch/twiki/bin/view/CMSPublic/CompOpsLocalSubmission#Site_customized_glideins_in_dept

and of course we should fully delete the deprecated text and rather paste in there
what is now in
https://twiki.cern.ch/twiki/bin/view/CMSPublic/CompOpsCustomizedGlideins

and then rewrite it in light of the current, improved understanding :-)

but I am slow... gimme time, even if I had my coffees already

**Comment by mmascher** [comment]
Hi both, this should do it. Requires this WMCore patch https://github.com/dmwm/WMCore/pull/6429

**Comment by mmascher** [comment]
Also, what about give the possibility to add Dns in addition to usernames to the local-users.txt files?

**Comment by bbockelm** [comment]
Hi,

I would prefer to call such a file `local-dns.txt`.

However, let's wait until the dust settles for VOMS groups and revisit?

The approach in the patches now look much better.  Who will be doing the testing?

Brian

**Comment by mmascher** [comment]
Sure, I agree, let's wait. Maybe even wait that somebody asks for it again.

For the testing we just had a CRAB3 release, but I guess we can ask @jmarra13  to apply the patches to the crabserver in the testbed to do some tests. Then I don't know who can volounteer to send tasks. We can talm about it at the SI meeting.

**Comment by bbockelm** [comment]
I can add a VOMS group to the `local-groups.txt` at Nebraska.  @jmarra13 - what roles do you have?  I'd prefer to have something somewhat limited to testers, not make a gift to German CMS users ;)

**Comment by belforte** [comment]
groups, not roles. grr...

we can use the old,obsolescent group /cms/integration
https://voms2.cern.ch:8443/voms/cms/group/edit.action?groupId=11

Marco is already a member. I will add Jadir and ... myself ? Brian ?

No, we do not want a file of DN's. There's no bloody point, given all the pain we suffer anyhow to 
maintain DN/username correspondence in SiteDB etc.
Let's not invent use cases out of thin air, simply because we can,
not anymore, please. please, please !!!! If someone has free time, please
help to streamline the SiteDB vs. DN thing.

**Comment by bbockelm** [comment]
Yeah - that's kind of my feeling on `local-dns.txt`; I'd like to treat that separately because I'd push on what the use case might be (not already covered by `local-users.txt`!).

`/cms/integration` is fine by me.  Please feel free to add me.

**Comment by mmascher** [comment]
Ok, agreed. Jadir, let us know if you install the patches on the testbed.
